### ISSUE TITLE
TypeScript broken on latest release

### ISSUE THREAD

**Issue by ChrisEdson** [Automattic/mongoose#14258]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

8.0.4

### Node.js version

20.x

### MongoDB server version

5.0.0

### Typescript version (if applicable)

5.3.3

### Description

I noticed that I'm no longer getting type checking for `find` queries in my code. 

The following should throw two errors, in both the new User constructor and the User.find:

```ts
import { Schema, model, connect } from "mongoose";

// 1. Create an interface representing a document in MongoDB.
interface IUser {
  name: string;
  email: string;
  avatar?: string;
}

// 2. Create a Schema corresponding to the document interface.
const userSchema = new Schema<IUser>({
  name: { type: String, required: true },
  email: { type: String, required: true },
  avatar: String,
});

// 3. Create a Model.
const User = model<IUser>("User", userSchema);

run().catch((err) => console.log(err));

async function run() {
  // 4. Connect to MongoDB
  await connect("mongodb://127.0.0.1:27017/test");

  const user1 = new User({
    name: "Bill",
    emal: "bill@initech.com",
    avaar: "https://i.imgur.com/dM7Thhn.png", // Should throw an error, since `avaar` is not in the schema
  });
  await user1.save();

  const user2 = await User.findOne({ emal: "" }).orFail(); // Should throw an error, since `emal` is not in the schema

  console.log(user2.emai); // Correctly throws an error as emai is not in the schema
}
```

I've reproduced this to show no TypeScript errors are thrown in this playground: https://playcode.io/1723905

### Steps to Reproduce

https://playcode.io/1723905

### Expected Behavior

Two TypeScript errors should be thrown.

**Comment by DarkOutX** [comment]
Seems like it's a critical bug. 
I don't get types too, and it happens coz IDE can't detect it due to too hard/big structure.
I get **+30C and 70-90%** of loading Ryzen 7 when open **files where mongoose imported**
Unable to use is at all, downgraded to 6x version, all fine

**Comment by vkarpov15** [comment]
This is expected behavior, you wouldn't get any type errors in Mongoose 7 or Mongoose 6 either for `await User.findOne({ emal: "" }).orFail()` or `avaar: "https://i.imgur.com/dM7Thhn.png"`. Mongoose doesn't do strict type checking in either case because 1) `await User.findOne({ emal: "" }).orFail()` is perfectly valid by default Mongoose won't strip out `emal` from the filter, 2) `avaar` will just get stripped out by default.

**Comment by ChrisEdson** [comment]
Thanks so much for getting back to me Valeri.

I believe we used to have this type checking behaviour, but it could have been back in Mongoose 5 (and when using `@types/mongoose`.

Can I ask in situation would the current behaviour be preferable?

At the very least, I'd love a feature where this is opt-in (perhaps on a per Schema or per Model basis).

On upgrading to from Mongoose 6 to 7, I've uncovered many many bugs where we have incorrectly typed filter arguments in our `Model.find`s.

I thought this was being picked up by TypeScript, but it appears that it isn't.
